parsec/core/fs/userfs/userfs.py

```diff
@@ -61,13 +63,96 @@
     FSWorkspaceInMaintenance,
     FSWorkspaceNotInMaintenance,
 )
+from parsec.core.fs.utils import is_file_manifest
 
 
 logger = get_logger()
 
 AnyEntryName = Union[EntryName, str]
 
 
+def _handle_maintenance_job_exception(exc, kind, workspace_id):
+    if exc.status in ("not_in_maintenance", "bad_encryption_revision"):
+        raise FSWorkspaceNotInMaintenance(f"{kind.capitalize()} job already finished") from exc
+
+    if exc.status == "not_allowed":
+        raise FSWorkspaceNoAccess(
+            f"Not allowed to do {kind} maintenance on workspace {workspace_id}: {exc}"
+        ) from exc
+
+    raise FSError(f"Cannot do {kind} maintenance on workspace {workspace_id}: {exc}") from exc
+
+
+class GarbageCollectionJob:
+    def __init__(self, backend_cmds, workspace_entry, device):
+        self.backend_cmds = backend_cmds
+        self.workspace_entry = workspace_entry
+        self.device = device
+        self._minimum_counter = 2
+        self._blocks_to_keep = {}
+        self.couter_by_versions = {}
+
+    async def do_one_batch(self, datetime_limit=None, size=100):
+        if not datetime_limit:
+            datetime_limit = pendulum_now().add(months=-6)
+        workspace_id = self.workspace_entry.id
+        try:
+            batch = await self.backend_cmds.vlob_maintenance_get_garbage_collection_batch(
+                workspace_id, size
+            )
+            total = 0
+            done = 0
+            done_batch = []
+            for vlob_id, version, datetime, blob in batch:
+                try:
+                    manifest = RemoteManifest.decrypt_verify_and_load(
+                        blob,
+                        key=self.workspace_entry.key,
+                        author_verify_key=self.device.verify_key,
+                        expected_author=self.device.device_id,
+                        expected_timestamp=datetime,
+                        expected_version=version,
+                        expected_id=vlob_id,
+                    )
+                except DataError as exc:
+                    raise FSError(f"Cannot decrypt vlob: {exc}") from exc
+
+                blocks_to_erase = set()
+                if is_file_manifest(manifest):
+                    try:
+                        _ = self._blocks_to_keep[vlob_id]
```

Comment:
`if vlob_id not in self._blocks_to_keep` ?

(et pourquoi pas un defaultdict ?)